File: PRODUCTION/PRODUCTION_2/backend/src/routes/verify.ts
import { Router } from 'express';
import { z } from 'zod';
import prisma from '../lib/prisma.js';

const router = Router();

// In-memory storage for demo
const verifications: any[] = [];

// Validation schema
const verifySchema = z.object({
  type: z.enum(['email', 'phone', 'identity', 'document']),
  value: z.string().min(1)
});

// Email verification logic
function verifyEmail(email: string) {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  const isValid = emailRegex.test(email);
  
  let riskScore = 100;
  if (email.includes('test') || email.includes('temp')) riskScore -= 20;
  if (email.endsWith('.ru') || email.endsWith('.cn')) riskScore -= 10;
  if (!email.includes('.')) riskScore -= 30;
  
  return {
    valid: isValid,
    riskScore: Math.max(0, Math.min(100, riskScore)),
    details: {
      format: isValid ? 'valid' : 'invalid',
      domain: email.split('@')[1] || 'unknown'
    }
  };
}

// Phone verification logic
function verifyPhone(phone: string) {
  const phoneRegex = /^\+?[1-9]\d{1,14}$/;
  const isValid = phoneRegex.test(phone.replace(/[\s-()]/g, ''));
  
  const riskScore = isValid ? 85 : 30;
  
  return {
    valid: isValid,
    riskScore,
    details: {
      format: isValid ? 'valid' : 'invalid',
      country: phone.startsWith('+1') ? 'US' : 'unknown'
    }
  };
}

// Identity verification logic
function verifyIdentity(name: string) {
  const isValid = name.length >= 2 && /^[a-zA-Z\s]+$/.test(name);
  
  const riskScore = isValid ? 75 : 20;
  
  return {
    valid: isValid,
    riskScore,
    details: {
      format: isValid ? 'valid' : 'invalid',
      length: name.length
    }
  };
}

// Document verification logic
function verifyDocument(docNumber: string) {
  const isValid = docNumber.length >= 5;
  
  const riskScore = isValid ? 80 : 25;
  
  return {
    valid: isValid,
    riskScore,
    details: {
      format: isValid ? 'valid' : 'invalid',
      length: docNumber.length
    }
  };
}

// Main verification endpoint
router.post('/', async (req, res) => {
  try {
    const { type, value } = verifySchema.parse(req.body);

    let result;
    switch (type) {
      case 'email':
        result = verifyEmail(value);
        break;
      case 'phone':
        result = verifyPhone(value);
        break;
      case 'identity':
        result = verifyIdentity(value);
        break;
      case 'document':
        result = verifyDocument(value);
        break;
      default:
        return res.status(400).json({ error: 'Invalid verification type' });
    }

    const verification = {
      id: crypto.randomUUID(),
      type,
      value,
      ...result,
      timestamp: new Date().toISOString()
    };

    // Try to save to database
    try {
      await prisma.verification.create({
        data: {
          type,
          inputData: { value },
          result: result,
          riskScore: result.riskScore,
          status: 'completed',
          completedAt: new Date(),
        },
      });
    } catch (dbError) {
      // Fallback to in-memory storage
      verifications.push(verification);
    }

    res.json({
      valid: result.valid,
      riskScore: result.riskScore,
      details: {
        type,
        value,
        timestamp: verification.timestamp,
        ...result.details
      }
    });
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: 'Validation error', details: error.errors });
    }
    console.error('Verification error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get verification by ID
router.get('/:id', async (req, res) => {
  const { id } = req.params;

  try {
    const verification = await prisma.verification.findUnique({
      where: { id },
    });

    if (!verification) {
      return res.status(404).json({ error: 'Verification not found' });
    }

    res.json(verification);
  } catch (dbError) {
    // Fallback to in-memory storage
    const verification = verifications.find(v => v.id === id);
    if (!verification) {
      return res.status(404).json({ error: 'Verification not found' });
    }
    res.json(verification);
  }
});

// List verifications
router.get('/', async (req, res) => {
  const limit = parseInt(req.query.limit as string) || 10;
  const offset = parseInt(req.query.offset as string) || 0;

  try {
    const results = await prisma.verification.findMany({
      take: limit,
      skip: offset,
      orderBy: { createdAt: 'desc' },
    });

    res.json({
      verifications: results,
      limit,
      offset
    });
  } catch (dbError) {
    // Fallback to in-memory storage
    const results = verifications.slice(offset, offset + limit);
    res.json({
      verifications: results,
      limit,
      offset
    });
  }
});

export default router;

File: PRODUCTION/PRODUCTION_3/backend/app/api/verify.py
"""
Verification API Routes
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, EmailStr
from typing import Literal, Optional
from datetime import datetime
import re
import uuid

router = APIRouter()

# In-memory storage
verifications_db = []

# Schemas
class VerificationRequest(BaseModel):
    type: Literal["email", "phone", "identity", "document"]
    value: str

class VerificationResponse(BaseModel):
    id: str
    type: str
    valid: bool
    risk_score: int
    details: dict
    timestamp: datetime

# Verification logic
def verify_email(email: str) -> dict:
    """Verify email address"""
    email_regex = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
    is_valid = bool(re.match(email_regex, email))
    
    risk_score = 100
    if 'test' in email.lower() or 'temp' in email.lower():
        risk_score -= 20
    if email.endswith('.ru') or email.endswith('.cn'):
        risk_score -= 10
    if not '.' in email:
        risk_score -= 30
    
    return {
        "valid": is_valid,
        "risk_score": max(0, min(100, risk_score)),
        "details": {
            "format": "valid" if is_valid else "invalid",
            "domain": email.split('@')[1] if '@' in email else "unknown",
            "disposable": "test" in email.lower() or "temp" in email.lower()
        }
    }

def verify_phone(phone: str) -> dict:
    """Verify phone number"""
    phone_clean = re.sub(r'[\s\-\(\)]', '', phone)
    phone_regex = r'^\+?[1-9]\d{1,14}$'
    is_valid = bool(re.match(phone_regex, phone_clean))
    
    risk_score = 85 if is_valid else 30
    
    return {
        "valid": is_valid,
        "risk_score": risk_score,
        "details": {
            "format": "valid" if is_valid else "invalid",
            "country": "US" if phone.startswith('+1') else "unknown",
            "type": "mobile" if is_valid else "unknown"
        }
    }

def verify_identity(name: str) -> dict:
    """Verify identity/name"""
    is_valid = len(name) >= 2 and bool(re.match(r'^[a-zA-Z\s]+$', name))
    
    risk_score = 75 if is_valid else 20
    
    return {
        "valid": is_valid,
        "risk_score": risk_score,
        "details": {
            "format": "valid" if is_valid else "invalid",
            "length": len(name),
            "parts": len(name.split()) if is_valid else 0
        }
    }

def verify_document(doc_number: str) -> dict:
    """Verify document number"""
    is_valid = len(doc_number) >= 5
    
    risk_score = 80 if is_valid else 25
    
    return {
        "valid": is_valid,
        "risk_score": risk_score,
        "details": {
            "format": "valid" if is_valid else "invalid",
            "length": len(doc_number),
            "type": "passport" if len(doc_number) >= 8 else "id"
        }
    }

# Routes
@router.post("/", response_model=VerificationResponse)
async def create_verification(request: VerificationRequest):
    """Create new verification"""
    
    # Perform verification based on type
    if request.type == "email":
        result = verify_email(request.value)
    elif request.type == "phone":
        result = verify_phone(request.value)
    elif request.type == "identity":
        result = verify_identity(request.value)
    elif request.type == "document":
        result = verify_document(request.value)
    else:
        raise HTTPException(status_code=400, detail="Invalid verification type")
    
    # Create verification record
    verification = {
        "id": str(uuid.uuid4()),
        "type": request.type,
        "value": request.value,
        "valid": result["valid"],
        "risk_score": result["risk_score"],
        "details": result["details"],
        "timestamp": datetime.utcnow()
    }
    
    verifications_db.append(verification)
    
    return VerificationResponse(**verification)

@router.get("/{verification_id}", response_model=VerificationResponse)
async def get_verification(verification_id: str):
    """Get verification by ID"""
    verification = next(
        (v for v in verifications_db if v["id"] == verification_id),
        None
    )
    
    if not verification:
        raise HTTPException(status_code=404, detail="Verification not found")
    
    return VerificationResponse(**verification)

@router.get("/", response_model=list[VerificationResponse])
async def list_verifications(limit: int = 10, offset: int = 0):
    """List verifications"""
    return [
        VerificationResponse(**v)
        for v in verifications_db[offset:offset + limit]
    ]

